File: java/ankit_kanyal/devnest_problems/NextGreaterElement.java
// link- https://devsnest.in/algo-challenges/next-greater-element

static int[] solve(int n, int[] nums){
//CODE HERE 
int ans[]=new int[n];
Arrays.fill(ans,-1);
Stack<Integer> stack=new Stack<>();

for(int i=0;i<2*n;i++){

    int index=i%n;
    while(!stack.empty() && nums[stack.peek()]<nums[index]){
        ans[stack.pop()]=nums[index];
    }
    stack.push(index);
}


return ans;
}

File: CPP/Sejal/NextGreaterElement.cpp
// https://devsnest.in/lms?id=47&slug=next-greater-element

vector<int> solve(int n, vector<int> nums){
//CODE HERE 
// code again
    int i;
    vector<int> ans(n, 0);
    stack<int> st;
    for (i = 2*n - 1; i >= 0; i--){
        while (!st.empty() && st.top() <= nums[i % n])
            st.pop();
        if (!st.empty()) {
            ans[i % n] = st.top();
        }
        else {
            ans[i % n] = -1;
        } 
        st.push(nums[i % n]);
    }
    return ans;
}
